File: riveredge-backend/src/apps/kuaiplm/models/sample_process.py
"""样品加工（R-15 #33）

通用单据：申请种类 + 附件类型由行业扩展 profile 驱动；
未启用电子包时仅「通用加工」与中性物料编码/版本。
"""

from tortoise import fields

from core.models.base import BaseModel


class SampleProcessApplication(BaseModel):
    """研发项目样品加工。"""

    class Meta:
        table = "apps_kuaiplm_sample_process_applications"
        table_description = "快研发 - 样品加工"
        unique_together = [("tenant_id", "application_code")]
        indexes = [
            ("tenant_id", "project_id"),
            ("tenant_id", "status"),
            ("tenant_id", "request_kind"),
            ("tenant_id", "material_code"),
            ("uuid",),
        ]

    id = fields.IntField(pk=True, description="主键")
    application_code = fields.CharField(max_length=50, description="申请单号")
    project_id = fields.IntField(description="研发项目ID")
    project_code = fields.CharField(max_length=50, description="项目代号快照")
    project_name = fields.CharField(max_length=200, description="项目名称快照")
    request_kind = fields.CharField(
        max_length=30,
        default="general",
        description="申请种类 code（来自 profile）",
    )
    title = fields.CharField(max_length=200, description="申请标题")
    material_code = fields.CharField(
        max_length=100, null=True, description="物料编码（行业包可改展示为 PCB 料号）"
    )
    material_version = fields.CharField(
        max_length=50, null=True, description="物料版本"
    )
    release_date = fields.DateField(null=True, description="资料发布日期")
    purpose = fields.TextField(null=True, description="申请原因/用途")
    status = fields.CharField(
        max_length=20,
        default="draft",
        description="draft/pending/approved/rejected/closed",
    )
    # [{attachment_type, file_uuid, file_name}]
    attachments = fields.JSONField(default=list, description="附件清单")
    remarks = fields.TextField(null=True, description="备注")
    submitted_at = fields.DatetimeField(null=True, description="提交时间")
    approved_at = fields.DatetimeField(null=True, description="审核通过时间")
    closed_at = fields.DatetimeField(null=True, description="关闭时间")
    deleted_at = fields.DatetimeField(null=True, description="软删除")

File: riveredge-frontend/src/apps/kuaiplm/services/sample-process.ts
/**
 * 样品加工申请 API（R-15 #33）
 * 种类/附件/字段文案由 /meta/form-profile 驱动（通用或电子行业包）。
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/sample-process-applications';

export type SampleProcessStatus = 'draft' | 'pending' | 'approved' | 'rejected' | 'closed';

export interface SampleProcessAttachment {
  attachment_type: string;
  file_uuid: string;
  file_name?: string | null;
}

export interface SampleProcessApplication {
  id: number;
  uuid: string;
  application_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  request_kind: string;
  title: string;
  material_code?: string | null;
  material_version?: string | null;
  release_date?: string | null;
  purpose?: string | null;
  status: SampleProcessStatus;
  attachments?: SampleProcessAttachment[];
  remarks?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  closed_at?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface SampleProcessPayload {
  project_id: number;
  request_kind: string;
  title: string;
  application_code?: string;
  material_code?: string | null;
  material_version?: string | null;
  release_date?: string | null;
  purpose?: string | null;
  attachments?: SampleProcessAttachment[];
  remarks?: string | null;
}

export interface SampleProcessFormProfileItem {
  code: string;
  label: string;
  sort?: number;
  active?: boolean;
}

export interface SampleProcessFormProfile {
  request_kinds: SampleProcessFormProfileItem[];
  attachment_types: SampleProcessFormProfileItem[];
  field_labels: Record<string, string>;
  validation_rules: Array<{
    when_kind_in?: string[];
    require?: string[];
    message?: string;
  }>;
}

function unwrapList(res: unknown): { items: SampleProcessApplication[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as SampleProcessApplication[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const sampleProcessApi = {
  formProfile: async () => {
    const res = await api.get(`${BASE}/meta/form-profile`);
    return res as SampleProcessFormProfile;
  },
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    project_id?: number;
    request_kind?: string;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as SampleProcessApplication;
  },
  create: async (data: SampleProcessPayload) => {
    const res = await api.post(BASE, data);
    return res as SampleProcessApplication;
  },
  update: async (id: number, data: Partial<SampleProcessPayload>) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as SampleProcessApplication;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as SampleProcessApplication;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as SampleProcessApplication;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as SampleProcessApplication;
  },
  close: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/close`);
    return res as SampleProcessApplication;
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};
